listen.go

```diff
@@ -149,78 +135,98 @@ func (l *listener) handleIncoming() {
 	defer wg.Done()
 
 	for {
-		maconn, err := l.Listener.Accept()
+		conn, err := l.Listener.Accept()
 		if err != nil {
 			if l.catcher.IsTemporary(err) {
 				continue
 			}
-
-			l.incoming <- connErr{err: err}
+			l.incoming <- connOrErr{err: err}
 			return
 		}
 
-		log.Debugf("listener %s got connection: %s <---> %s", l, maconn.LocalMultiaddr(), maconn.RemoteMultiaddr())
-
-		if l.filters != nil && l.filters.AddrBlocked(maconn.RemoteMultiaddr()) {
-			log.Debugf("blocked connection from %s", maconn.RemoteMultiaddr())
-			maconn.Close()
+		if l.filters != nil && l.filters.AddrBlocked(conn.RemoteMultiaddr()) {
+			log.Debugf("blocked connection from %s", conn.RemoteMultiaddr())
+			conn.Close()
 			continue
 		}
-		// If we have a wrapper func, wrap this conn
-		if l.wrapper != nil {
-			maconn = l.wrapper(maconn)
-		}
+
+		log.Debugf("listener %s got connection: %s <---> %s", l, conn.LocalMultiaddr(), conn.RemoteMultiaddr())
 
 		wg.Add(1)
 		go func() {
 			defer wg.Done()
-			if l.protec != nil {
-				pc, err := l.protec.Protect(maconn)
+
+			ctx, cancel := context.WithTimeout(l.ctx, ConnAcceptTimeout)
+			defer cancel()
+
+			done := make(chan struct{})
+			go func() {
+				defer close(done)
+
+				if l.protec != nil {
+					pc, err := l.protec.Protect(conn)
+					if err != nil {
+						conn.Close()
+						log.Warning("protector failed: ", err)
+						return
+					}
+					conn = pc
+				}
+
+				// If we have a wrapper func, wrap this conn
+				if l.wrapper != nil {
+					conn = l.wrapper(conn)
+				}
+
+				if _, _, err := l.mux.Negotiate(conn); err != nil {
+					log.Warning("incoming conn: negotiation of crypto protocol failed: ", err)
+					conn.Close()
+					return
+				}
+
+				c, err := newSingleConn(ctx, l.local, "", l.privk, conn, l.streamMuxer, true)
 				if err != nil {
-					maconn.Close()
-					log.Warning("protector failed: ", err)
+					log.Warning("connection setup failed: ", err)
+					conn.Close()
+					return
 				}
-				maconn = pc
-			}
 
-			maconn.SetReadDeadline(time.Now().Add(NegotiateReadTimeout))
-			_, _, err = l.mux.Negotiate(maconn)
-			if err != nil {
-				log.Warning("incoming conn: negotiation of crypto protocol failed: ", err)
-				maconn.Close()
+				l.incoming <- connOrErr{conn: c}
+			}()
+
+			select {
+			case <-ctx.Done():
```

Comment:
There's a race here. We can return the connection and then close it.

Reply:
I pushed a fix for that.